### DISCUSSION TITLE
functions for targets (switch case targets)

### DISCUSSION THREAD

**Discussion by alita-moore** [statelyai/xstate#4442] [Ideas]
I am working on a task where a service returns some insight into the context to the machine. The machine then decides which state it should transition to. The logic is as followings:
```ts
const fileTypeSwitch = (context, event) => {
  const fileType = event.data.fileType;
  switch (fileType) {
    case "raw-image":
      return States.UPLOADING_RAW_IMAGE;
    case "document":
      return States.UPLOADING_DOCUMENT;
    case "unsupported":
      return States.UNSUPPORTED_FILE_TYPE;
    default:
      // @ts-expect-error -- if this doesn't error a case is missed above
      fileType.toString();
      return States.ERROR;
  }
}
```

Basically, this fileTypeSwitch returns the next state that it should go to. Ideally, I could plus this into the "target" attribute and be good to go, but it seems that's not supported.

Instead I had to create the following:
```ts
invoke: {
        src: InvokableServices.IDENTIFY_FILE_TYPE,
        onDone: [
          States.ERROR,
          States.UPLOADING_RAW_IMAGE,
          States.UPLOADING_DOCUMENT,
          States.UNSUPPORTED_FILE_TYPE,
        ].map(state => ({
          target: state,
          cond: fileTypeSwitch(state),
        })),
        onError: x.onError,
      }
```

where it now injects which state it is looking for and then sets the conditional accordingly. The implementation of this is just 
```ts
(state) => fileSwitchCase(context, event) === state
```

Having the target be able to be determined by a function would simplify this code greatly.

**Comment by Andarist** [comment]
> Having the target be able to be determined by a function would simplify this code greatly.

True - but it would also hinder our static analysis possibilities which are super important. We know about this struggle though and discussed this a bunch of times so we might be able figure out a solution to that at some point.

**Reply by alita-moore** [reply]
How do you mean, exactly? My code is type safe, if that helps. Like the type itself is the same as if you were to provide the objects directly. I.e. Like [{ target: States.whatever, cond: () > boolean},...]

**Reply by davidkpiano** [reply]
Basically, if we were to allow a function that resolves to a target, you could do:

```ts
target: () => {
  return Math.random() < 0.5 ? 'there' : 'elsewhere';
}
```

We wouldn't be able to visualize nor serialize that easily.

**Reply by alita-moore** [reply]
I see, ya that's a tough problem. Internally I was building a helper function that would just output the relevant json output that is standard right now. I mean in my post it does basically that and the code works as expected. Can't you all offer that helper as a work around here?

**Reply by Andarist** [reply]
We can do something about this but we need more time to flesh out the idea properly and to make sure that the solution we come up with checks all the required checkboxes for us. It's also a matter of priority. Right now we focus on different things - so while we recognize the need, we can't act on it immediately. I'd rather have a properly thought-through solution rather than a one taken hastily.
